File: code/jvm/http-pipeline/src/main/java/pt/isel/pipeline/authentication/RequestTokenProcessor.java
package pt.isel.pipeline.authentication;

import org.springframework.stereotype.Component;
import pt.isel.domain.users.AuthenticatedUser;
import pt.isel.services.users.UserService;

@Component
public class RequestTokenProcessor {
    public static final String SCHEME = "bearer";
    private final UserService usersService;

    public RequestTokenProcessor(UserService usersService) {
        this.usersService = usersService;
    }

    public AuthenticatedUser processAuthorizationHeaderValue(String authorizationValue) {
        if (authorizationValue == null || authorizationValue.isBlank()) return null;

        String[] parts = authorizationValue.trim().split("\\s+");
        if (parts.length != 2 || !parts[0].equalsIgnoreCase(SCHEME)) return null;

        String token = parts[1];
        var user = usersService.getUserByToken(token);
        return user != null ? new AuthenticatedUser(user, token) : null;
    }
}

File: code/jvm/http-pipeline/src/main/kotlin/pt/isel/pipeline/authentication/RequestTokenProcessor.kt
package pt.isel.pipeline.authentication

import org.springframework.stereotype.Component
import pt.isel.domain.users.AuthenticatedUser
import pt.isel.services.users.UserService

@Component
class RequestTokenProcessor(
    private val usersService: UserService,
) {
    fun processAuthorizationHeaderValue(authorizationValue: String?): AuthenticatedUser? {
        if (authorizationValue.isNullOrBlank()) return null

        val parts = authorizationValue.split(' ').filter { it.isNotBlank() }
        if (parts.size != 2 || !parts[0].equals(SCHEME, ignoreCase = true)) return null

        val token = parts[1]
        return usersService.getUserByToken(token)?.let { AuthenticatedUser(it, token) }
    }

    companion object {
        const val SCHEME = "bearer"
    }
}
